iceberg/iceberg-catalog/src/main/java/org/apache/iceberg/hive/HiveActor.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.hive;
+
+import java.util.List;
+import java.util.Map;
+import org.apache.hadoop.hive.metastore.api.Database;
+import org.apache.hadoop.hive.metastore.api.LockRequest;
+import org.apache.hadoop.hive.metastore.api.LockResponse;
+import org.apache.hadoop.hive.metastore.api.ShowLocksRequest;
+import org.apache.hadoop.hive.metastore.api.ShowLocksResponse;
+import org.apache.hadoop.hive.metastore.api.Table;
+import org.apache.iceberg.TableMetadata;
+import org.apache.iceberg.catalog.Namespace;
+import org.apache.thrift.TException;
+
+/**
+ * Acts as the Hive client for the HiveCatalog benefit.
+ */
+public interface HiveActor {
```

Comment:
To me, the `HiveActor` is not necessary. The embedded Hive can be implemented by setting `hive.metastore.uris` to an empty string during creating the `HiveCatalog`.
The new interface could introduce some burden if we want to port the Iceberg changes to Hive codebase in the future.

Reply:
I beg to differ. My raw understanding of the code, starting with the HiveCatalog, requires a pool of IMetaStoreClient. The HiveMetaStore (the server part) itself can be instantiated as being local (isMetaStoreRemote=false) but this does not seem to make the client (IMetaStoreClient) any different; the method isMetaStoreRemote() is called only to help during autumn/authz from what I gather. This means that at least any call requires serializing/deserializing Thrift messages and go through the loopback interface (127.0.0.1) for no reason which is a cost we don't have to incur being embedded. Furthermore, this new interface does provide some isolation from HMS and instead of the whole IMetaStoreClient interface, we got a narrow set of identified behaviors required for the Iceberg Catalog only (not the whole legacy that HMS carries).

Reply:
Finally got what you said with some external help ;-) : if we are to set the THRIFT_URIS to blank in the config, the client created (HiveMetaStoreClient actor) checks MetastoreConf.isEmbeddedMetaStore(...) and calls callEmbeddedMetastore(...) which is not based on a Thrift client but a genuine IHMSHandler. By deriving HiveCatalog and overriding the method createPool() with a config modified to blank THRIFT_URIS, we now have a way to create embedded HiveCatalog. I wish I learnt that earlier...